packages/@react-aria/toast/src/useToast.ts

```diff
@@ -10,85 +10,102 @@
+  descriptionProps: DOMAttributes,
+  /** Props for the toast close button. */
+  closeButtonProps: AriaButtonProps
 }
 
-export function useToast(props: ToastAriaProps, state: ToastState): ToastAria {
+/**
+ * Provides the behavior and accessibility implementation for a toast component.
+ * Toasts display brief, temporary notifications of actions, errors, or other events in an application.
+ */
+export function useToast<T>(props: AriaToastProps<T>, state: ToastState<T>, ref: RefObject<FocusableElement>): ToastAria {
   let {
-    toastKey,
-    onAction,
-    onClose,
-    shouldCloseOnAction,
+    key,
     timer,
-    variant
-  } = props;
-  let {
-    onRemove
-  } = state;
-  let stringFormatter = useLocalizedStringFormatter(intlMessages);
-  let domProps = filterDOMProps(props);
+    timeout,
+    animation
+  } = props.toast;
 
-  const handleAction = (...args) => {
-    if (onAction) {
-      onAction(...args);
+  useEffect(() => {
+    if (!timer) {
+      return;
     }
 
-    if (shouldCloseOnAction) {
-      onClose && onClose(...args);
-      onRemove && onRemove(toastKey);
-    }
-  };
+    timer.reset(timeout);
+    return () => {
+      timer.pause();
+    };
+  }, [timer, timeout]);
 
-  let iconProps = variant ? {'aria-label': stringFormatter.format(variant)} : {};
+  // Restore focus to the toast container on unmount.
+  // If there are no more toasts, the container will be unmounted
+  // and will restore focus to wherever focus was before the user
+  // focused the toast region.
+  let focusOnUnmount = useRef(null);
+  useLayoutEffect(() => {
+    let container = ref.current.closest('[role=region]') as HTMLElement;
+    return () => {
+      if (container && container.contains(document.activeElement)) {
+        // Focus must be delayed for focus ring to appear, but we can't wait
+        // until useEffect cleanup to check if focus was inside the container.
+        focusOnUnmount.current = container;
+      }
+    };
+  }, [ref]);
 
-  let pauseTimer = () => {
-    timer && timer.pause();
-  };
+  // eslint-disable-next-line
+  useEffect(() => {
+    return () => {
+      if (focusOnUnmount.current) {
+        focusOnUnmount.current.focus();
+      }
+    };
+  }, [ref]);
 
-  let resumeTimer = () => {
-    timer && timer.resume();
-  };
-
-  let {hoverProps} = useHover({
-    onHoverStart: pauseTimer,
-    onHoverEnd: resumeTimer
-  });
-
-  let {focusProps} = useFocus({
-    onFocus: pauseTimer,
-    onBlur: resumeTimer
-  });
+  let titleId = useId();
+  let descriptionId = useSlotId();
+  let stringFormatter = useLocalizedStringFormatter(intlMessages);
 
   return {
-    toastProps: mergeProps(domProps, {
-      ...hoverProps,
-      role: 'alert'
-    }),
-    iconProps,
-    actionButtonProps: {
-      ...focusProps,
-      onPress: handleAction
+    toastProps: {
+      role: 'alert',
+      'aria-label': props['aria-label'],
+      'aria-labelledby': props['aria-labelledby'] || titleId,
```

Comment:
Maybe. TBH I'm not sure if this is even working. The goal was to prevent screen readers from announcing the close button after reading the content of the toast (by pointing `aria-labelledby` directly to the title), but in my testing VoiceOver at least still reads everything. So this might not actually do anything at all. Further testing in other screen readers needed. Will handle later.